### PR TITLE
Add `badge` circle style

### LINKED ISSUE #3: Badge circle style
[Issuehunt badges](https://issuehunt.io/r/sindresorhus/DockProgress/issues/3)

Would be nice to have a badge circle style like what Chrome has for downloads:

screen shot 2018-02-25 at 21 13 55

> [jonivr](https://issuehunt.io/u/jonivr) earned $80.00 by resolving this issue!
> - Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/DockProgress/) to discover more funded issues.
> - Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on Issuehunt to raise funds.

### PR DESCRIPTION
**Pull request by JoniVR**
Hi,

I implemented #3 and updated the example to loop through it as well. 
Let me know if there are any things you'd like me to change or fix. 

The `badgeLabel` only allows for 3 characters though, after that it will truncate because text resizing in `CATextLayer` is very hard (I tried for several hours but couldn't find a solution) and not really useful anyways imo.

Best regards,
Joni

**edit:**
I'm sorry about the number of commits, I wasn't fully satisfied with how it looked. 
Feel free to squash if you feel like that looks cleaner.

Fixes #3

### CODE REVIEW

**Comment by sindresorhus** [comment]
Nice work! It already looks pretty good :)

**Comment by sindresorhus** [comment]
I noticed the circle doesn't start exactly in the middle:

screen shot 2019-01-17 at 15 38 05

File: Sources/DockProgress/util.swift [hunk 1]
```diff
@@ -106,3 +110,31 @@ extension NSBezierPath {
...
+final class VerticallyCenteredTextLayer: CATextLayer {
+
+	convenience init(frame rect: CGRect, center: CGPoint) {
+		self.init()
+		frame = rect
+		frame.center = center
+		alignmentMode = .center
+		truncationMode = .end
+		font = NSFont(name: "HelveticaNeue-Bold", size: 0)
+		fontSize = 25
```
Comment (sindresorhus) [inline Sources/DockProgress/util.swift]: These font props are two specific for such a generic class and should be set at callsite instead.

File: Sources/DockProgress/util.swift [hunk 2]
```diff
@@ -106,3 +110,31 @@ extension NSBezierPath {
...
+		contentsScale = NSScreen.main!.backingScaleFactor
```
Comment (sindresorhus) [inline Sources/DockProgress/util.swift]:
```suggestion
		contentsScale = NSScreen.main?.backingScaleFactor ?? 2
```

The current code will fail if the system has no screen.

File: Sources/DockProgress/util.swift [hunk 3]
```diff
@@ -41,7 +41,11 @@ final class ProgressCircleShapeLayer: CAShapeLayer {
 		self.init()
 		fillColor = nil
 		lineCap = .round
+		position = center
 		path = NSBezierPath.progressCircle(radius: radius, center: center).cgPath
+		if let boundingBox = path?.boundingBox {
+			bounds = boundingBox
+		}
```
Comment (sindresorhus) [inline Sources/DockProgress/util.swift]: Can you explain these changes and why they are needed?

File: Sources/DockProgress/DockProgress.swift [hunk 4]
```diff
@@ -32,6 +32,7 @@ public final class DockProgress {
 		case bar
 		/// TODO: Make `color` optional when https://github.com/apple/swift-evolution/blob/master/proposals/0155-normalize-enum-case-representation.md is shipping in Swift
 		case circle(radius: Double, color: NSColor)
+		case badge(color: NSColor, badgeLabel: String)
```
Comment (sindresorhus) [inline Sources/DockProgress/DockProgress.swift]: Having the `badeLabel` as a `String` means it's impossible to update it during its run, which is the whole point of having it in the first place. We could either make it a closure that is expected to return a string or add a new optional property to the DockProgress instance that is only used for the `badge` style. Thoughts on which one to pick?

**Comment by sindresorhus** [comment]
> The badgeLabel only allows for 3 characters though, after that it will truncate because text resizing in CATextLayer is very hard (I tried for several hours but couldn't find a solution) and not really useful anyways imo.

We need resizing as even for 3 characters it looks very cramped. We could hard-code the sizes for each characters though.

**Comment by sindresorhus** [comment]
I also think we should just accept a number instead of a string. I can't imagine using this for anything other than a number and having a number means we can have some smart behavior for it. For example, when the number passes `999`, we can switch it to `1K` style (means 1000). I think Apple has an API to format numbers like this automatically.

So:

- `999` => `999`
- `1000` => `1K`
- `1100` => `1K`
- `2000` => `2K`
- `10000` => `9K+`

**Comment by sindresorhus** [comment]
Here's a comparison with the Chrome badge:

screen shot 2019-01-17 at 15 34 05

I like yours better, but I think it could be a tiny bit smaller, but not as small as the Chrome one.

Also notice how the Chrome one has a slight shadow under the badge. We need that too.

File: Sources/DockProgress/util.swift [hunk 5]
```diff
@@ -106,3 +110,29 @@ extension NSBezierPath {
...
 }
+
+final class VerticallyCenteredTextLayer: CATextLayer {
+	convenience init(frame rect: CGRect, center: CGPoint) {
+		self.init()
+		frame = rect
+		frame.center = center
+		alignmentMode = .center
+		truncationMode = .end
+		font = NSFont(name: "HelveticaNeue-Bold", size: 0)
```
Comment (sindresorhus) [inline Sources/DockProgress/util.swift]: Make this an extension on `NSFont` for ease of use.

Like:

```swift
extension NSFont {
	static let helveticaNeueBold = NSFont(name: "HelveticaNeue-Bold", size: 0)
```

File: Sources/DockProgress/util.swift [hunk 6]
```diff
@@ -106,3 +110,29 @@ extension NSBezierPath {
...
 	}
 }
+
+final class VerticallyCenteredTextLayer: CATextLayer {
+	convenience init(frame rect: CGRect, center: CGPoint) {
+		self.init()
+		frame = rect
+		frame.center = center
+		alignmentMode = .center
+		truncationMode = .end
```
Comment (sindresorhus) [inline Sources/DockProgress/util.swift]: This is also unrelated to vertically centering.

File: Sources/DockProgress/DockProgress.swift [hunk 7]
```diff
@@ -97,4 +100,36 @@ public final class DockProgress {
...
+		let radius = dstRect.width / 4.5
+		let newCenter = CGPoint(x: dstRect.maxX - radius - 4, y: dstRect.minY + radius + 4)
+
+		// Background
+		let badge = ProgressCircleShapeLayer(radius: Double(radius), center: newCenter)
+		badge.fillColor = CGColor(red: 0.94, green: 0.96, blue: 1, alpha: 1)
+
+		// Progress circle
+		let lineWidth: CGFloat = 6
+		let progressCircle = ProgressCircleShapeLayer(radius: Double(radius) - Double(lineWidth / 2), center: newCenter)
```
Comment (sindresorhus) [inline Sources/DockProgress/DockProgress.swift]: Can you move `Double(radius) - Double(lineWidth / 2)` into a variable?

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: Sources/DockProgress/DockProgress.swift (hunk 4 above)
Comment (JoniVR) [reply Sources/DockProgress/DockProgress.swift]: You're right. I'd go with closure since that seems like a better approach to me. I'll look into it.

File: Sources/DockProgress/DockProgress.swift (hunk 7 above)
Comment (JoniVR) [reply Sources/DockProgress/DockProgress.swift]: I'll look into it 🙂

File: Sources/DockProgress/util.swift (hunk 3 above)
Comment (JoniVR) [reply Sources/DockProgress/util.swift]: As far as I can remember, the `ProgressCircleShapeLayer` had no bounds or position values, this made it very hard to align the `CATextLayer` with it. They don't break anything else, they just add some extra correct values to it. 

I figured I'd assign these values so they are correct and can be used to center the `CATextLayer` and give it the same frame so that the dimensions match too. I had a really hard time getting the `CATextLayer` centered horizontally and vertically correct with the badge itself.

If you know of a better way to go about this, or don't agree with this I'd be happy to change it up.

File: Sources/DockProgress/util.swift (hunk 6 above)
Comment (JoniVR) [reply Sources/DockProgress/util.swift]: Correct, I just thought it was cleaner to put it inside the subclass, I could either move it somewhere else (where it probably doesn't add much value) or rename the subclass?

**Comment by JoniVR** [comment]
I'll look into fixing these things when I get some more time, I'm currently also working on a Caprine PR and a project for a client, so my time is going to be divided between these three projects.

File: Sources/DockProgress/DockProgress.swift (hunk 4 above)
Comment (sindresorhus) [reply Sources/DockProgress/DockProgress.swift]: Alright. Closure it is.

File: Sources/DockProgress/util.swift (hunk 6 above)
Comment (sindresorhus) [reply Sources/DockProgress/util.swift]: The subclass is useful for other situations too though. I wouldn't rename it. Just set the props where it's used.

**Comment by sindresorhus** [comment]
@JoniVR No worries. There's no rush on this. Take as long as you need.

**Comment by JoniVR** [comment]
> Here's a comparison with the Chrome badge:
> 
> screen shot 2019-01-17 at 15 34 05
> 
> I like yours better, but I think it could be a tiny bit smaller, but not as small as the Chrome one.
> 
> Also notice how the Chrome one has a slight shadow under the badge. We need that too.

screenshot 2019-01-17 at 13 26 06

How about this size?
(font still needs to be adjusted)

**Comment by JoniVR** [comment]
`// TODO:`

- [x] fix circle doesn't start exactly in the middle
- [x] character resizing (`Formatter`)
- [x] use closure for `badgeLabel`
- [x] add `badge` shadow
- [x] move contentScale back to `VerticallyCenteredTextLayer`

File: Sources/DockProgress/util.swift [hunk 8]
```diff
@@ -116,11 +119,6 @@ final class VerticallyCenteredTextLayer: CATextLayer {
 		self.init()
 		frame = rect
 		frame.center = center
-		alignmentMode = .center
-		truncationMode = .end
-		font = NSFont(name: "HelveticaNeue-Bold", size: 0)
-		fontSize = 25
-		contentsScale = NSScreen.main!.backingScaleFactor
```
Comment (sindresorhus) [inline Sources/DockProgress/util.swift]: This one (line 123) could stay here. It's generally useful.

**Comment by sindresorhus** [comment]
> How about this size?

👍

File: Sources/DockProgress/DockProgress.swift (hunk 4 above)
Comment (JoniVR) [reply Sources/DockProgress/DockProgress.swift]: Hmmm now that I've implemented it with I closure I'm thinking it might not be the most user-friendly thing.. Like you'd have to call it like this (for example) right?
```swift
.badge(color: .systemBlue, badgeLabel: {
				return "\(Int(DockProgress.progressValue * 100))"
			})
```

**Comment by JoniVR** [comment]
@sindresorhus 	

Is it just me or is the other one *slightly* off as well? 
screenshot 2019-01-19 at 01 13 56

screenshot 2019-01-19 at 01 19 54

*I set the `lineCap = .square` on this one to make sure it wasn't the `linecap`.*

It would make sense since I'm basically reusing that part and it probably just looks worse because of the smaller radius, right?

File: Sources/DockProgress/DockProgress.swift [hunk 9]
```diff
@@ -58,6 +59,8 @@ public final class DockProgress {
 				self.drawProgressBar(dstRect)
 			case let .circle(radius, color):
 				self.drawProgressCircle(dstRect, radius: radius, color: color)
+			case let .badge(color, badgeValue):
+				self.drawProgressBadge(dstRect, badgeLabel: badgeValue(), color: color)
```
Comment (sindresorhus) [inline Sources/DockProgress/DockProgress.swift]: Nitpick, but the argument order should be in the same order as `case badge(color: NSColor, badgeValue: () -> Int)`

File: Example/AppDelegate.swift [hunk 10]
```diff
@@ -12,14 +12,28 @@ final class AppDelegate: NSObject, NSApplicationDelegate {
 	func applicationDidFinishLaunching(_ notification: Notification) {
 		borrowIconFrom(app: "Photos")
 
-		var lastStyleWasBar = true
+		let styles: [DockProgress.ProgressStyle] = [
+			.bar,
+			.circle(radius: 58, color: .systemPink),
+			.badge(color: .systemBlue, badgeValue: {
+				return Int(DockProgress.progressValue * 100)
```
Comment (sindresorhus) [inline Example/AppDelegate.swift]: The `return` is not needed here.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
> Is it just me or is the other one slightly off as well?

Not just you.

> It would make sense since I'm basically reusing that part and it probably just looks worse because of the smaller radius, right?

Probably, yes. We still need to figure out what's causing it though.

**Comment by JoniVR** [comment]
I wasn't able to use the NumberFormatter Apple provided, I don't think it does what we want (unless I missed it). 

I implemented a custom `Int` variable `shortStringRepresentation` for it instead. 
I've also made the font scale dynamically on the badge.
Let me know if you disagree with something or if there's still an issue somewhere 😄

I'll look into the progress circle not starting in the middle, but I've looked into it before and your code looked fine to me, must be something small.

**edit:**
I figured it out in the badge case, simply using `.butt` instead of `.square` for the `lineCap` property seems to fix it as far as I can tell, doesn't fix the other one, but you barely notice that one anyways, I don't think anything else is causing the issue.. unless I'm missing something.

Here's a gif of it in action:
ezgif com-video-to-gif

File: Sources/DockProgress/util.swift [hunk 11]
```diff
@@ -134,4 +156,18 @@ final class VerticallyCenteredTextLayer: CATextLayer {
...
+			charCount = string.length
+		}
+		switch charCount {
+		case 1: return 30.0
+		case 2: return 23.0
+		case 3: return 20.0
+		case 4: return 15.0
+		default: return 0.0
+		}
+	}
```
Comment (sindresorhus) [inline Sources/DockProgress/util.swift]: This should not be an extension to `VerticallyCenteredTextLayer`. It's not related at all to vertically centering.

File: Sources/DockProgress/util.swift [hunk 12]
```diff
@@ -134,4 +156,18 @@ final class VerticallyCenteredTextLayer: CATextLayer {
 		super.draw(in: context)
 		context.restoreGState()
 	}
+
+	var dynamicFontSizeForBadge: CGFloat {
+		var charCount = 0
+		if let string = self.string as? NSString {
+			charCount = string.length
+		}
```
Comment (sindresorhus) [inline Sources/DockProgress/util.swift]: Why not use `string.count`?

File: Sources/DockProgress/util.swift [hunk 13]
```diff
@@ -59,6 +59,28 @@ final class ProgressCircleShapeLayer: CAShapeLayer {
...
+		let signum = self.signum()
+		let number = abs(self)
+		if number < 1000 {
+			return "\(self)"
+		} else if number < 10000 {
+			return "\(signum * Int(number/1000))K"
+		} else {
+			return "\(signum * 9)K+"
+		}
+	}
```
Comment (sindresorhus) [inline Sources/DockProgress/util.swift]: The logic here is too specific to be a generic utility function. Better to make it just a private method on the DockProgress class instead.

**Comment by sindresorhus** [comment]
Can you document it in the readme?

File: Sources/DockProgress/util.swift [hunk 14]
```diff
@@ -106,3 +135,39 @@ extension NSBezierPath {
...
+
+	var dynamicFontSizeForBadge: CGFloat {
+		var charCount = 0
+		if let string = self.string as? NSString {
+			charCount = string.length
+		}
+		switch charCount {
+		case 1: return 30.0
+		case 2: return 23.0
+		case 3: return 20.0
```
Comment (sindresorhus) [inline Sources/DockProgress/util.swift]: This feels a bit cramped when it's `9K+`. Let's use `19`.

File: Sources/DockProgress/util.swift [hunk 15]
```diff
@@ -55,13 +59,38 @@ final class ProgressCircleShapeLayer: CAShapeLayer {
...
+	*/
+	var shortStringRepresentation: String {
+		let signum = self.signum()
+		let number = abs(self)
+		if number < 1000 {
+			return "\(self)"
+		} else if number < 10000 {
+			return "\(signum * Int(number/1000))K"
+		} else {
+			return "\(signum * 9)K+"
```
Comment (sindresorhus) [inline Sources/DockProgress/util.swift]: I just realized that the `k` should be lowercase. https://en.wikipedia.org/wiki/Kilo-

File: Sources/DockProgress/util.swift [hunk 16]
```diff
@@ -55,13 +59,39 @@ final class ProgressCircleShapeLayer: CAShapeLayer {
...
+	/**
+	```
+	999 => 999
+	1000 => 1K
+	1100 => 1K
+	2000 => 2K
+	10000 => 9K+
+	```
+	*/
+	var shortStringRepresentation: String {
```
Comment (sindresorhus) [inline Sources/DockProgress/util.swift]: This needs a better name. I think `kiloShortStringFromInt()` would work.

File: Sources/DockProgress/util.swift [hunk 17]
```diff
@@ -41,7 +41,11 @@ final class ProgressCircleShapeLayer: CAShapeLayer {
 		self.init()
```
Comment (sindresorhus) [inline Sources/DockProgress/util.swift:41]: This init could be simplified to this:

```swift
convenience init(radius: Double, center: CGPoint) {
	self.init()
	fillColor = nil
	lineCap = .round
	position = center
	strokeEnd = 0

	let cgPath = NSBezierPath.progressCircle(radius: radius, center: center).cgPath
	path = cgPath
	bounds = cgPath.boundingBox
}
```

File: Sources/DockProgress/DockProgress.swift [hunk 18]
```diff
@@ -97,4 +100,46 @@ public final class DockProgress {
...
+		badge.fillColor = CGColor(red: 0.94, green: 0.96, blue: 1, alpha: 1)
+		badge.shadowColor = .black
+		badge.shadowOpacity = 0.3
+		badge.masksToBounds = false
+		badge.shadowOffset = CGSize(width: -1, height: 1)
+		badge.shadowPath = badge.path
+
+		// Progress circle
+		let lineWidth: CGFloat = 6
+		let innerRadius = Double(radius) - Double(lineWidth / 2)
```
Comment (sindresorhus) [inline Sources/DockProgress/DockProgress.swift]: It's more efficient to convert to `Double` where it's used (line 124) instead of for each variable.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
> I figured it out in the badge case, simply using .butt instead of .square for the lineCap property seems to fix it as far as I can tell

👍

File: Sources/DockProgress/util.swift (hunk 12 above)
Comment (JoniVR) [reply Sources/DockProgress/util.swift]: Because, apparently, `NSString` uses `.length` instead of `.count` for whatever reason, it confused me at first too.

Comment (sindresorhus) [reply Sources/DockProgress/util.swift]: The question was, why not use `String`?

Comment (JoniVR) [reply Sources/DockProgress/util.swift]: I am about to push a commit which refactors the code and uses String instead ;)

**Comment by JoniVR** [comment]
I'll start updating the documentation now, let me know if there's anything else that needs changing.

File: Sources/DockProgress/DockProgress.swift [hunk 19]
```diff
@@ -97,4 +100,86 @@ public final class DockProgress {
...
+			return "\(sign * Int(absNumber / 1000))k"
+		} else {
+			return "\(sign * 9)k+"
+		}
+	}
+
+	private static func scaledBadgeFontSize(text: String) -> CGFloat {
+		let charCount = text.count
+
+		switch charCount {
```
Comment (sindresorhus) [inline Sources/DockProgress/DockProgress.swift]: Just use `text.count` directly here.

**Comment by JoniVR** [comment]
Should the documentation contain the same code as the current code example?
I'm going to record a gif at the same dimensions as the other ones as well.

**Comment by sindresorhus** [comment]
> Should the documentation contain the same code as the current code example?

Yes

> I'm going to record a gif at the same dimensions as the other ones as well.

👌

**Review by sindresorhus — APPROVED** [review]

**Comment by sindresorhus** [comment]
Looks great now. Thanks for this awesome contribution, @JoniVR! 🎉

**Comment by JoniVR** [comment]
@sindresorhus 
It was a pleasure, and I learned a lot! 
Thank you for all your work! 😁
